### PR TITLE
Changes for upcoming remote observing API

### PR DESCRIPTION
**Pull request by joshjriley**

### CODE REVIEW

**Comment by joshjriley** [comment]
Do we want to get rid of instructions to generate a helpdesk ticket to "Notify Keck of your intent to connect remotely"?  Part of that is to send along all the info about their remote/home station.  If so, should we have them enter that info on the web with their SSH key upload?  Would require a little bit of dev time to expand the webform and figure out where in the DB to put that info.

**Comment by joshjriley** [comment]
The test_all portion currently has API 'account' param hard-coded to 'kcwi1'.  The API requires an instrument account which we don't necessarily have when they run '--test'.  We could change the API to accept no account, but Jeff suggested instead we get rid of the connection test to all 10 vm-instrs and instead just have them test a real instr account during their test window.

**Comment by joshjriley** [comment]
I think there is a prior bug when firewall password is entered incorrectly which results in a strange failure.  Code flounders around a bit doing other stuff after trying to exit app.  This pushes the warning that the "API key is not set" off the screen. 
 We might want to fix this b/c this might come up a lot when existing users don't follow instructions to add api_key and we change the fw pass on them.

**Comment by joshwalawender** [comment]
I haven't dug in to the details of the code yet (soon I hope), but I think for the failure you mention in the most recent comment, we should add a warning for anyone not using the API.  

Another way to distinguish old setups (i.e. current users) who haven't adopted the new API yet would be to deprecate the current username for the firewall password and use something else in the future.  Then we just check if they're using the old username in the config file and error out for anyone still using that firewall user name.

File: keck_vnc_launcher.py [hunk 1]
```diff
@@ -18,16 +19,18 @@
...
 import warnings
 import yaml
 
 ## Import local modules
 import soundplay
 
 
-__version__ = '1.2.4'
+## Module vars
+__version__ = '1.3.0'
```
Comment (joshwalawender) [inline keck_vnc_launcher.py]: I suggest going to v2.0.0, this is a big change.

File: keck_vnc_launcher.py [hunk 2]
```diff
@@ -91,6 +94,10 @@ def create_parser():
     ## add options
     parser.add_argument("-c", "--config", dest="config", type=str,
         help="Path to local configuration file.")
+    parser.add_argument("-vncserver", type=str,
+        help="Name of VNC server to connect to.  Takes precedence over all.")
+    parser.add_argument( '-vncports', nargs='+', type=str,
```
Comment (joshwalawender) [inline keck_vnc_launcher.py]: I suggest double dashes to follow the conventions already in use, so `--vncserver` and `--vncports`

File: keck_vnc_launcher.py [hunk 3]
```diff
@@ -447,6 +458,16 @@ def start(self):
                 self.log.error('Exiting program.')
                 self.exit_app()
 
+        ##---------------------------------------------------------------------
+        ## Get connect info from API
+        if self.api_key:
+            self.get_api_data(self.api_key, self.args.account)
+            if self.api_data is None:
+                if self.firewall_defined is True:
+                    self.log.info('Firewall info detected in config.  Trying ate method.')
```
Comment (joshwalawender) [inline keck_vnc_launcher.py]: "ate method"?

File: keck_vnc_launcher.py (hunk 2 above)
Comment (joshjriley) [reply keck_vnc_launcher.py]: whoops, just an oversight.  will fix.

File: keck_vnc_launcher.py (hunk 3 above)
Comment (joshjriley) [reply keck_vnc_launcher.py]: "alternate".  Must have made a mistake editing text.  will fix.

File: keck_vnc_launcher.py [hunk 4]
```diff
@@ -1067,17 +1179,7 @@ def validate_ssh_key(self):
 
         self.ssh_key_valid = False
         cmd = 'whoami'
-#         server = self.servers_to_try[0]
-
-        # Find a server in the servers_to_try list with is not an svncserverN
-        server = None
-        for s in self.servers_to_try:
```
Comment (joshwalawender) [inline keck_vnc_launcher.py]: Why is this block being removed?  I think it is still necessary (though only in the case where the API is not functioning of course).

Comment (joshjriley) [reply keck_vnc_launcher.py]: I can’t even remember editing that function, so I think it was a rogue edit that didn’t get caught in my testing b/c it still works.  I’ll put it back to what it was.

**Comment by joshjriley** [comment]
> I haven't dug in to the details of the code yet (soon I hope), but I think for the failure you mention in the most recent comment, we should add a warning for anyone not using the API.
> 
> Another way to distinguish old setups (i.e. current users) who haven't adopted the new API yet would be to deprecate the current username for the firewall password and use something else in the future. Then we just check if they're using the old username in the config file and error out for anyone still using that firewall user name.

**Comment by joshjriley** [comment]
mistakenly close pull request

File: keck_vnc_launcher.py [hunk 5]
```diff
@@ -824,6 +863,79 @@ def get_vncviewer_properties(self):
...
 
 
+    def get_api_data(self, api_key, account):
+        '''Get data from API which contains all info needed to connect.'''
+        self.api_data = None
+
+        #form API url and get data
+        url = f'{KRO_API}?key={self.api_key}'
+        if account is not False: url += f'&account={self.args.account}'
+        self.log.debug(f'Calling API with URL: {url}')
```
Comment (joshwalawender) [inline keck_vnc_launcher.py:879]: We should also log what responses we got from the API.  If we don't want to put some info in the log for security reasons, we should just have a log line which says that we got that info.

**Comment by joshwalawender** [comment]
This is looking pretty good, but we haven't closed the loop on the VNC password.  We either need to print that to the screen so the user can type it in, or we need to decide to do away with it.

File: keck_vnc_launcher.py [hunk 6]
```diff
@@ -1114,23 +1229,38 @@ def get_vnc_server(self, account, instrument):
...
+            self.log.info("Using VNC server defined on command line")
+            vncserver = self.args.vncserver
 
-            if data is not None and ' ' not in data:
-                vncserver = data
-                self.log.info(f"Got VNC server: '{vncserver}'")
-                break
+        #API Route
+        elif self.api_data:
+            self.log.info(f"Determining VNC server for '{self.args.account}' (via API)")
```
Comment (joshwalawender) [inline keck_vnc_launcher.py:1245]: Need to log what server we got, similar to line 1262 below.

Comment (joshjriley) [reply keck_vnc_launcher.py:1245]: ok done

File: keck_vnc_launcher.py (hunk 5 above)
Comment (joshjriley) [reply keck_vnc_launcher.py:879]: ok done.  Logging API return code and if successfully parsed.  I'm thinking we shouldn't log the actual response b/c of all the sensitive info in it.  Easy to add if we change mind.

**Comment by joshjriley** [comment]
> This is looking pretty good, but we haven't closed the loop on the VNC password. We either need to print that to the screen so the user can type it in, or we need to decide to do away with it.

I added call to view_connection_info at the very end just before re-displaying menu.  Added  border around it to make it extra visible.

**Pull request merged**